src/allocation_engine/README.md

```diff
@@ -0,0 +1,170 @@
+# Design
+
+## Allocation engine
+
+This implementation uses an interval tree that is specialised for allocation of
+memory-mapped I/O and port I/O address space. The fields of the strctures
+defined will have semantic meaning for this context (e.g node state to indicate
+if a node in the tree is assigned or not to a device).
+
+We offer three options for placing a memory slot in the managed address space:
+
+1. `LastMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the end of the managed address space.
+2. `FirstMatch` -> When using this allocation policy the allocator will try to
+insert the range described by the constraint at the first available position
+starting from the begining of the managed address space.
+3. `ExactMatch` -> When using this allocation policy the allocator will try to
+insert the range at the exact position described by the constraint, otherwise
+it will return an error.
+
+```rust
```

Comment:
This kind of docs tend to get outdated really fast because we'll forget to update the docs when updating the code. If it doesn't bring much value, I would suggest we remove them from the document, especially since you are already describing the contents very well in the previous text block.